frame/system/src/offchain.rs

```diff
@@ -17,351 +17,487 @@
+	) -> Result<(), ()> {
+		let xt = T::Extrinsic::new(call.into(), signature).ok_or(())?;
+		sp_io::offchain::submit_transaction(xt.encode())
+	}
+
+	pub fn submit_unsigned_transaction(
+		call: <T as SendTransactionTypes<LocalCall>>::OverarchingCall,
+	) -> Result<(), ()> {
+		SubmitTransaction::<T, LocalCall>::submit_transaction(call, None)
+	}
 }
 
-/// A trait responsible for signing a payload using given account.
+/// Provides an implementation for signing transaction payloads
 ///
-/// This trait is usually going to represent a local public key
-/// that has ability to sign arbitrary `Payloads`.
+/// Keys used for signing are defined when instantiating the signer object.
+/// Signing can be done using:
 ///
-/// NOTE: Most likely you don't need to implement this trait manually.
-/// It has a blanket implementation for all `RuntimeAppPublic` types,
-/// so it's enough to pass an application-specific crypto type.
+/// - All supported keys in the keystore
+/// - Any of the supported keys in the keystore
+/// - A list of provided keys
 ///
-/// To easily create `SignedTransaction`s have a look at the
-/// [`TransactionSubmitter`] type.
-pub trait Signer<Public, Signature> {
-	/// Sign any encodable payload with given account and produce a signature.
-	///
-	/// Returns `Some` if signing succeeded and `None` in case the `account` couldn't
-	/// be used (for instance we couldn't convert it to required application specific crypto).
-	fn sign<Payload: Encode>(public: Public, payload: &Payload) -> Option<Signature>;
+/// The signer is then able to:
+/// - Submit a unsigned transaction with a signed payload
+/// - Submit a signed transaction
+pub struct Signer<T: SigningTypes, C: AppCrypto<T::Public, T::Signature>, X = ForAny> {
+	accounts: Option<Vec<T::Public>>,
+	_phantom: sp_std::marker::PhantomData<(X, C)>,
 }
 
-/// A `Signer` implementation for any `AppPublic` type.
-///
-/// This implementation additionally supports conversion to/from multi-signature/multi-signer
-/// wrappers.
-/// If the wrapped crypto doesn't match `AppPublic`s crypto `None` is returned.
-impl<Public, Signature, TAnyAppPublic> Signer<Public, Signature> for TAnyAppPublic where
-	TAnyAppPublic: RuntimeAppPublic
-		+ AppPublic
-		+ From<<TAnyAppPublic as AppPublic>::Generic>,
-	<TAnyAppPublic as RuntimeAppPublic>::Signature: AppSignature,
-	Signature: From<
-		<<TAnyAppPublic as RuntimeAppPublic>::Signature as AppSignature>::Generic
-	>,
-	Public: TryInto<<TAnyAppPublic as AppPublic>::Generic>
-{
-	fn sign<Payload: Encode>(public: Public, raw_payload: &Payload) -> Option<Signature> {
-		raw_payload.using_encoded(|payload| {
-			let public = public.try_into().ok()?;
-			TAnyAppPublic::from(public).sign(&payload)
-				.map(
-					<<TAnyAppPublic as RuntimeAppPublic>::Signature as AppSignature>
-					 ::Generic::from
-				)
-				.map(Signature::from)
-		})
+impl<T: SigningTypes, C: AppCrypto<T::Public, T::Signature>, X> Default for Signer<T, C, X> {
```

Comment:
is this needed? Both option and PhranotomData have a default impl.

Reply:
it is needed because of `X` I suppose.

Reply:
You're right in the fact that we are implementing `Default` twice here where X could be either `ForAll` or `ForAny`